### DISCUSSION TITLE
Empty regex causes metrics inside pods to be not scraped

### DISCUSSION THREAD

**Discussion by flokli** [grafana-cold-storage/agent#437] [Q&A] [answered]
I deployed the agent according to https://github.com/grafana/agent/#getting-started, which essentially just uses `https://github.com/grafana/agent/blob/v0.12.0/production/kubernetes/agent.yaml` under the hood.

For some reason, the "empty regex" relabel_config at

https://github.com/grafana/agent/blob/e9bd098ffd09930015173008ffb254018e7333c4/production/kubernetes/agent.yaml#L51-L54

seems to be problematic - with it, I couldn't scrape pod workloads exporting their own metrics on a `/metrics` endpoint.

What's that line supposed to be doing in first place?

**Comment by rfratto** [comment, marked as answer]
Hi! It's [looking for a label named `name`](https://github.com/grafana/agent/blob/e9bd098ffd09930015173008ffb254018e7333c4/docs/getting-started.md#kubernetes-install-script). Do your pods have that?

**Reply by flokli** [reply]
Huh, 3 is somewhat unexpected, and back when I looked at it, https://github.com/grafana/agent/commit/2d42f428186d7181c8744308f48a9364edc18589 didn't yet land.

Is there a reason the `name` label is required?

**Reply by flokli** [reply]
So, the pods do not have a `name` label, but an `app` label.

**Reply by rfratto** [reply]
No, there's no reason you couldn't change the rules here. We copied the rules from Grafana's [prometheus-ksonnet](https://github.com/grafana/jsonnet-libs/blob/b1d8f68fda371984ba5fa927adbac7c84d0db50e/prometheus-ksonnet/lib/prometheus-config.libsonnet#L77-L82), which I believe was deployed in an environment where every pod was expected to have a name label, and anything else was a resource that should've been ignored.

We're definitely looking to improve these rules, since I suspect the set we're using now isn't useful for most of the users deploying the Agent on their k8s clusters.

**Reply by flokli** [reply]
I want to stick as much as possible with (your) upstream config, cause I don't want do revisit these every time there's an update. Dropping that condition shouldn't probably be too much of a problem.

On that note - I opened https://github.com/grafana/agent/issues/439, which would also save a lot of headaches, due to how config/secrets are handled.

**Reply by rfratto** [reply]
Dropping the condition wouldn't break anything, but we're pretty concerned with making the Kubernetes install scrape _more_ things by default in between releases. We don't want someone to update their Agent install and suddenly start sending so many metrics their Grafana Cloud bill goes up, which would be awful. 

Over the next few months you'll see some new things coming in that change our Kubernetes support: 

1. An Operator in the style of Prometheus Operator, which we'll shift to using as our recommended installation for Kubernetes 
2. Helm charts to install the Operator, so we can move away from the (frankly) super hacky install scripts

I'm open to discussing #439, but just warning you that it's not a priority from our side since soon-ish we're going to abandon the bash script approach for something more customizable. (Also it was designed with the intent of having one command to copy-and-paste when going through the Grafana Cloud setup, but that's a discussion we can have over on the issue directly)

**Reply by flokli** [reply]
Is there a document describing some of the design ideas of that operator?

> Helm charts to install the Operator, so we can move away from the (frankly) super hacky install scripts

Please still provide manifests people can trivially `kubectl apply`, refer from kustomize, or import into terraform via things like `k2tf`.

Regarding https://github.com/grafana/agent/issues/439, let's move that discussion bit over there.

**Reply by rfratto** [reply]
> Please still provide manifests people can trivially kubectl apply, refer from kustomize, or import into terraform via things like k2tf.

We're planning on keeping the Jsonnet configs around, at least, but I'm inclined to remove the bash scripts. Running bash scripts tend to make people uncomfortable, and it's hard to come up with sane plain YAML that would work for most users. The defaults we've been running with tend to be really expensive when you run it in a cloud environment. 

As-as now, the plan is still for Operator Helm charts for the very-easy-get-started with customization, and Jsonnet for power users.

**Reply by flokli** [reply]
I assume with more work going into the grafana agent operator, this stance has shifted a bit more to "let's use the `ServiceMonitor` from the prometheus CRD to describe what to monitor", over more complicated custom grafana-agent configs?

**Reply by flokli** [reply]
(from https://github.com/grafana/agent/issues/734#issuecomment-878413337)

**Reply by rfratto** [reply]
> I assume with more work going into the grafana agent operator, this stance has shifted a bit more to "let's use the ServiceMonitor from the prometheus CRD to describe what to monitor", over more complicated custom grafana-agent configs?

Yeah, generally I feel like that has the highest chance of ensuring a uniform experience and generally just making it easier to configure.
